atom/browser/api/atom_api_session.cc

```diff
@@ -214,54 +214,82 @@ std::map<uint32_t, v8::Global<v8::Object>> g_sessions;
 void RunCallbackInUI(const base::Callback<void()>& callback) {
   base::PostTaskWithTraits(FROM_HERE, {BrowserThread::UI}, callback);
 }
+
 template <typename... T>
 void RunCallbackInUI(const base::Callback<void(T...)>& callback, T... result) {
   base::PostTaskWithTraits(FROM_HERE, {BrowserThread::UI},
                            base::BindOnce(callback, result...));
 }
 
+void RejectPromise(atom::util::Promise promise, int error) {
+  std::string errmsg =
+      "Failed to size failed with error number: " + std::to_string(error);
+  promise.RejectWithErrorMessage(errmsg);
+}
+
+void RejectPromiseInUI(atom::util::Promise promise, int error) {
+  base::PostTaskWithTraits(
+      FROM_HERE, {BrowserThread::UI},
+      base::BindOnce(RejectPromise, std::move(promise), error));
+}
+
+template <typename T>
+void ResolvePromise(atom::util::Promise promise, T result) {
+  promise.Resolve(result);
+}
+
+template <typename T>
+void ResolvePromiseInUI(atom::util::Promise promise, T result) {
+  base::PostTaskWithTraits(
+      FROM_HERE, {BrowserThread::UI},
+      base::BindOnce(ResolvePromise<T>, std::move(promise), result));
+}
+
 // Callback of HttpCache::GetBackend.
 void OnGetBackend(disk_cache::Backend** backend_ptr,
                   Session::CacheAction action,
-                  const net::CompletionCallback& callback,
+                  const atom::util::CopyablePromise& promise,
                   int result) {
   if (result != net::OK) {
-    RunCallbackInUI(callback, result);
+    RejectPromiseInUI(promise.GetPromise(), result);
   } else if (backend_ptr && *backend_ptr) {
     if (action == Session::CacheAction::CLEAR) {
       (*backend_ptr)
```

Comment:
Should check the result of [DoomAllEntries](https://chromium.googlesource.com/chromium/src/+/master/net/disk_cache/disk_cache.h#217) for `net::ERR_IO_PENDING` . 

```
auto success = (*backend_ptr)->DoomAllEntries(
              base::BindOnce(&ResolveOrRejectPromiseInUI, promise));

if (success != net::ERR_IO_PENDING) {
  ResolveOrRejectPromiseInUI(success, std::move(promise.GetPromise()))
}
```


Also it is better to have a helper function to resolve or reject according to net error codes.

```
void ResolveOrRejectPromiseInUI(int net_error, ...) {
  if (net_error != net::OK) {
    RejectPromiseInUI(net::ErrorToString(net_error), std::move(promise))
  } else {
    ResolvePromiseInUI(std::move(promise));
  }  
} 
```